source/extensions/common/tap/tap.h

```diff
@@ -0,0 +1,124 @@
+#pragma once
+
+#include "envoy/common/pure.h"
+#include "envoy/data/tap/v2alpha/wrapper.pb.h"
+#include "envoy/http/header_map.h"
+#include "envoy/service/tap/v2alpha/common.pb.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace Common {
+namespace Tap {
+
+/**
+ * Sink for sending tap messages.
+ */
+class Sink {
+public:
+  virtual ~Sink() = default;
+
+  /**
+   * Send a fully buffered trace to the sink.
+   * @param trace supplies the trace to send. The trace message is a discrete trace message (as
+   *        opposed to a portion of a larger trace that should be aggregated).
+   */
+  virtual void
+  submitBufferedTrace(std::shared_ptr<envoy::data::tap::v2alpha::BufferedTraceWrapper> trace) PURE;
+};
+
+/**
+ * Generic configuration for a tap extension (filter, transport socket, etc.).
+ */
+class ExtensionConfig {
+public:
+  virtual ~ExtensionConfig() = default;
+
+  /**
+   * @return the ID to use for admin extension configuration tracking (if applicable).
+   */
+  virtual const std::string& adminId() PURE;
+
+  /**
+   * Clear any active tap configuration.
+   */
+  virtual void clearTapConfig() PURE;
+
+  /**
+   * Install a new tap configuration.
+   * @param proto_config supplies the generic tap config to install. Not all configuration fields
+   *        may be applicable to an extension (e.g. HTTP fields). The extension is free to fail
+   *        the configuration load via exception if it wishes.
+   * @param admin_streamer supplies the singleton admin sink to use for output if the configuration
+   *        specifies that output type. May not be used if the configuration does not specify
+   *        admin output. May be nullptr if admin is not used to supply the config.
+   */
+  virtual void newTapConfig(envoy::service::tap::v2alpha::TapConfig&& proto_config,
+                            Sink* admin_streamer) PURE;
+};
+
+class Matcher;
+using MatcherPtr = std::unique_ptr<Matcher>;
+
+/**
+ * Base class for all tap matchers.
+ *
+ * A high level note on the design of tap matching which is different from other matching in Envoy
+ * due to a requirement to support streaming matching (match as new data arrives versus
+ * calculating the match given all available data at once).
+ * - The matching system is composed of a constant matching configuration. This is essentially
+ *   a tree of matchers given logical AND, OR, NOT, etc.
+ * - A per-stream/request matching status must be kept in order to compute interim match status.
+ * - In order to make this computationally efficient, the matching tree is kept in a vector, with
```

Comment:
I think it would be a good idea to explain how the match tree is mapped to a vector here. Is this a root-leaf path, or maybe a heap structure or maybe a flattened N-ary tree, etc.?

Reply:
I don't know the technical name. I guess it's a flattened N-ary tree. If you tell me the name I will update it or I will just call it a "flattened n-ary tree."

Reply:
I think the correct term would be "preorder traversal flattening of N-arary tree", I think you put the parent node down before the descendents, right?

Reply:
OK. Can you approve and I will change this comment in a follow up? I have quite a few changes stacked up behind this one and I would rather not wait for another merge/CI round.